src/corefx/System.Globalization.Native/holders.h

```diff
@@ -0,0 +1,103 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+#include <unicode/ucal.h>
+#include <unicode/uenum.h>
+#include <unicode/udatpg.h>
+#include <unicode/udat.h>
+#include <unicode/unum.h>
+#include <unicode/uldnames.h>
+#include <unicode/ures.h>
+
+// IcuHolder is a template that can manage the lifetime of a raw pointer to ensure that it is cleaned up at the correct
+// time.  The general usage pattern is to aquire some ICU resource via an _open call, then construct a holder using the
+// pointer and UErrorCode to manage the lifetime.  When the holder goes out of scope, the coresponding close method is
+// called on the pointer.
+template <typename T, typename Closer>
+class IcuHolder
+{
+  public:
+    IcuHolder(T* p, UErrorCode err)
+    {
+        m_p = U_SUCCESS(err) ? p : nullptr;
+    }
+
+    ~IcuHolder()
+    {
+        if (m_p != nullptr)
+        {
+            Closer()(m_p);
+        }
+    }
+
+  private:
+    T* m_p;
+    IcuHolder(const IcuHolder&) = delete;
+    IcuHolder operator=(const IcuHolder&) = delete;
+};
+
+struct UCalendarCloser
+{
+    void operator()(UCalendar* pCal) const
+    {
+        ucal_close(pCal);
+    }
+};
+
+struct UEnumerationCloser
+{
+    void operator()(UEnumeration* pEnum) const
+    {
+        uenum_close(pEnum);
+    }
+};
+
+struct UDateTimePatternGeneratorCloser
+{
+    void operator()(UDateTimePatternGenerator* pGenerator) const
+    {
+        udatpg_close(pGenerator);
+    }
+};
+
+struct UDateFormatCloser
+{
+    void operator()(UDateFormat* pDateFormat) const
+    {
+        udat_close(pDateFormat);
+    }
+};
+
+struct UNumberFormatCloser
+{
+    void operator()(UNumberFormat* pNumberFormat) const
+    {
+        udat_close(pNumberFormat);
+    }
+};
+
+struct ULocaleDisplayNamesCloser
+{
+    void operator()(ULocaleDisplayNames* pLocaleDisplayNames) const
+    {
+        uldn_close(pLocaleDisplayNames);
+    }
+};
+
+struct UResourceBundleCloser
+{
+    void operator()(UResourceBundle* pResourceBundle) const
+    {
+        ures_close(pResourceBundle);
+    }
+};
+
+typedef IcuHolder<UCalendar, UCalendarCloser> UCalendarHolder;
+typedef IcuHolder<UEnumeration, UEnumerationCloser> UEnumerationHolder;
+typedef IcuHolder<UDateTimePatternGenerator, UDateTimePatternGeneratorCloser> UDateTimePatternGeneratorHolder;
+typedef IcuHolder<UDateFormat, UDateFormatCloser> UDateFormatHolder;
+typedef IcuHolder<UNumberFormat, UNumberFormatCloser> UNumberFormatHolder;
+typedef IcuHolder<ULocaleDisplayNames, ULocaleDisplayNamesCloser> ULocaleDisplayNamesHolder;
+typedef IcuHolder<UResourceBundle, UResourceBundleCloser> UResourceBundleHolder;
```

Comment:
FWIW, you could cut down on the ceremony by just making the closers overloads of an IcuClose function. That would  eliminate the extra template arg on IcuHolder, which you could then use directly as IcuHolder<T> without the typedefs.